src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,765 @@
+///    pop_offset. As a result, the pop_offset can never overtake the push_offset.
+/// 6. If the length header is zero, the data in the buffer may be valid but if it is valid, it is locked and should
+/// not be modified or read.
+/// 7. If the 16 bytes pointed to by pop_offset are zero, then there is either another ongoing pop or there is no valid
+/// data in the buffer.
+pub struct ConcurrentRingBuffer {
+    // Indexes the first empty byte where buffers can be enqueued.
+    push_offset: *mut usize,
+    // Indexes the first buffer that can be popped.
+    pop_offset: *mut usize,
+    // Underlying buffer.
+    buffer: raw_array::RawArray<u8>,
+    /// Is the underlying memory managed by this module?
+    is_managed: bool,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+/// Associated functions.
+impl ConcurrentRingBuffer {
+    /// Creates a ring buffer.
+    #[allow(unused)]
+    pub fn new(capacity: usize) -> Result<Self, Fail> {
+        // Check if capacity is invalid.
+        if capacity == 0 {
+            let cause: String = format!("invalid capacity (capacity={})", capacity);
+            error!("new(): {}", &cause);
+            return Err(Fail::new(libc::EINVAL, &cause));
+        }
+
+        let push_offset: *mut usize = Self::alloc::<usize>()?;
+        unsafe { *push_offset = 0 };
+
+        let pop_offset: *mut usize = Self::alloc::<usize>()?;
+        unsafe { *pop_offset = 0 };
+
+        let me: Self = Self {
+            pop_offset,
+            push_offset,
+            buffer: raw_array::RawArray::<u8>::new(capacity)?,
+            is_managed: true,
+        };
+
+        // Initialize the first header to 0.
+        me.write_header(0, 0);
+
+        Ok(me)
+    }
+
+    /// Returns the effective capacity of the target ring buffer in bytes.
+    #[allow(unused)]
+    pub fn capacity(&self) -> usize {
+        self.buffer.capacity()
+    }
+
+    #[allow(unused)]
+    pub fn remaining_capacity(&self) -> usize {
+        let push_offset: usize = peek(self.push_offset);
+        let pop_offset: usize = peek(self.pop_offset);
+        //println!("{:?} {:?}", push_offset, pop_offset);
+        self.available_space(push_offset, pop_offset)
+    }
+
+    /// Attempts to insert a buffer of [len] bytes into the ring buffer.
+    pub fn try_push(&self, buf: &[u8]) -> Result<usize, Fail> {
+        let len: usize = buf.len();
+        if len == 0 {
+            return Err(Fail::new(libc::EINVAL, "Buffer must be non-zero length"));
+        }
+        // reserve_space will allocate space for the header.
+        if let Some(push_offset) = self.reserve_space(len) {
+            debug_assert!(push_offset % 2 == 0);
```

Comment:
We might want to replace 2 with HEADER_SIZE

Reply:
Indeed.